### PR TITLE
Make build settings parsing optionally take into account any associated xcconfig files

### PR DESCRIPTION
**Pull request by Ruenzuo**
There was a TODO in `AbstractTarget.resolved_build_setting` to implement the necessary logic to take in consideration any associated xcconfig files to targets or projects. While implementing this I realised that adding support for `$(inherited)` is basically a must for this addition to be really useful. I'm not sure if the inheritance rules are documented anywhere (besides the [awesome docs](https://pewpewthespells.com/blog/xcconfig_guide.html#BuildSettingInheritance) provided by @samdmarshall), so this logic is based on what I could see in Xcode 8.3 (check [sample project](https://github.com/Ruenzuo/xcodeproj_parse_xcconfig_settings/tree/master) for reference).

This PR is related to https://github.com/CocoaPods/Xcodeproj/issues/180

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,9 @@
 
 ##### Enhancements
 
-* None.  
+* Make build settings parsing optionally take into account any associated xcconfig files.  
+  [Renzo Crisóstomo](https://github.com/ruenzuo)
+  [#476](https://github.com/CocoaPods/Xcodeproj/pull/476)
```
Comment (segiddins) [inline CHANGELOG.md]: This should link to #180

File: lib/xcodeproj/project/object/build_configuration.rb [hunk 2]
```diff
@@ -68,10 +68,34 @@ def type
...
+        # file present and resolving inheritance between them
+        #
+        # @param [String] key
+        #        the key of the build setting.
+        #
+        # @return [String] The value of the build setting
+        #
+        def resolve_build_setting(key)
+          return build_settings[key] if base_configuration_reference.nil?
+          @config ||= Xcodeproj::Config.new(File.new(base_configuration_reference.real_path)).to_hash
```
Comment (segiddins) [inline lib/xcodeproj/project/object/build_configuration.rb]: Maybe extract this into its own method?

File: lib/xcodeproj/project/object/build_configuration.rb [hunk 3]
```diff
@@ -68,10 +68,34 @@ def type
...
+          return @config[key] if build_settings[key].nil?
+          return expand_build_setting(build_settings[key], @config[key]) if build_settings[key].include? '$(inherited)'
```
Comment (segiddins) [inline lib/xcodeproj/project/object/build_configuration.rb]: I believe `${inherited}` also works

**Comment by segiddins** [comment]
This also needs tests for OTHER_LDFLAGS, which I believe we have separate code paths for, as well as xcconfigs that use conditional settings

File: lib/xcodeproj/project/object/build_configuration.rb (hunk 3 above)
Comment (Ruenzuo) [inline lib/xcodeproj/project/object/build_configuration.rb]: Didn't know that, I'll give it a try.

File: lib/xcodeproj/project/object/build_configuration.rb (hunk 2 above)
Comment (Ruenzuo) [inline lib/xcodeproj/project/object/build_configuration.rb]: Sure.

**Comment by Ruenzuo** [comment]
Completely forgot that conditional settings existed, guess that shows how much I use those 😅. I'll do the required changed to support that, conditional assignment based on configuration sounds rather important for correctness here.

**Comment by Ruenzuo** [comment]
@segiddins I think I implemented all requested changes, with the exception of supporting xcconfig files that use conditional settings. Apparently only `sdk` and `arch` conditions are widely used, but `resolved_build_setting` method returns values [grouped by configuration](https://github.com/CocoaPods/Xcodeproj/blob/master/lib/xcodeproj/project/object/native_target.rb#L44-L45) so I don't know how to implement this without changing the API, which I'm not sure if it's something desired.

Most of the new behaviour is opt-in, with the exception of resolving inherited values. Should I also hide this behind another parameter / method?

File: lib/xcodeproj/constants.rb [hunk 4]
```diff
@@ -382,5 +382,12 @@ module Constants
     # @return [Array] The extensions which are associated with header files.
     #
     HEADER_FILES_EXTENSIONS = %w(.h .hh .hpp .ipp .tpp .hxx .def).freeze
+
+    # @return [Array] The keywords Xcode use to identify a build setting can
+    #                 inherit values from a previous precedence level
+    INHERITED_KEYWORDS = %w(
```
Comment (segiddins) [inline lib/xcodeproj/constants.rb]: @samdmarshall do you know if this encompasses all of the ways of spelling `inherited` in build settings?

Comment (samdmarshall) [inline lib/xcodeproj/constants.rb]: yeah i believe so

**Comment by orta** [comment]
👍

**Comment by dnkoutso** [comment]
@DanToml @orta any objections on merging this? Seems 👍

**Comment by orta** [comment]
Yeah - go for it

**Pull request merged**